### DISCUSSION TITLE
Print Spread Factor, and Power Consumption

### DISCUSSION THREAD

**Discussion by Junaid-Q** [mcci-catena/arduino-lorawan#203] [General]
Dear Mr. Terry Moore (@terrillmoore ),

I am using LMIC code for experiment purposes and would like to print the spread factor and power consumed for each payload sent to the Server. I do wanna look at both parameters i,e SF and Power consumption on Serial console on Arduino IDE. Would you please provide me with your kind instruction on how to do that? 

`LMIC_setLinkCheckMode(0);`
`LMIC_setDrTxpow(DR_SF7,14);`

I would like to change the value from 14 dBi to 20 dBi and DR_SF7 to DR_SF12 and do wanna see the result on Arduino Console.

Looking forward to hearing from you
Thank you

Best,

![image](https://user-images.githubusercontent.com/12583457/181465784-ad9e1139-4139-449b-9371-495ce332b81a.png)

**Comment by terrillmoore** [comment]
Hi, sure. My recommendation is to capture this info during EV_TXSTART processing, and then print it later when it's not time critical (e.g., during EV_JOIN_TXCOMPLETE).  (Printing takes time and the LMIC has hard-real-time paths.)

With respect to Tx power, there is some tricky handoff between the LMIC and the radio driver, I'll have to look up the exact correct variable to capture. Which version of LMIC are you using?

**Comment by Junaid-Q** [comment]
> EV_TXSTART

Dear Mr. Terry Moore,

Thanks for your prompt response. What I understood from your recommendation is to capture SPREAD_FACTOR information in the main code (**ttn-otaa-feather-us915-dht22.ino**) file, or do we have to define it in the **LMIC.C** file? 

      `case EV_TXSTART:
            Serial.println(F("EV_TXSTART"));
            break;
        case SPREAD_FACTOR:      // <=================
            break;
        case EV_TXCANCELED:
            Serial.println(F("EV_TXCANCELED"));
            break;
        case EV_RXSTART:
            /* do not print anything -- it wrecks timing */
            break;
        case EV_JOIN_TXCOMPLETE:
            Serial.println(F("SPREAD_FACTOR: Uplink Packet sent with Spread Factor"));    // <=================
            Serial.println(F("EV_JOIN_TXCOMPLETE: no JoinAccept"));
            break;` 

I added those two lines but it gives me an error that "SPREAD_FACTOR" was not declared in this scope. For sure it must show this error because we didn't define it in the code. Would you be kind enough to please let me know where and how can we define the statement for the Spread Factor? 

______________
For **Powe consumption**, I kindly looking forward to hearing from you, when you get a chance in your free time. I tried myself but couldn't solve the issue. 

I am using the latest version of **LMIC [4.1.1]**

Thank you



I kindly need the result like this **SAMPLE SNAPSHOT** 
![image](https://user-images.githubusercontent.com/12583457/181585203-00174314-9500-43b4-92c0-8961d62687b6.png)

**Comment by terrillmoore** [comment]
Sorry, I was not clear. While processing the event `EV_TXSTART`, the code should cache the values of:

* `LMIC.rps`: this has the spreading factor and other information.  To get the spreading factor, you use `getSf()` from `lorabase.h`, passing in the `rps` value grabbed from `LMIC.rps`.

* `LMIC.radio_txpow`: this has the requested transmit power, in dB.

Please refer to doc/RadioDriver.md for more info.

**Reply by terrillmoore** [reply]
By the way, please note that the names of these fields will change in LMIC v5. If this is a concern, I suggest you wrap them in subroutines; the contents of the LMIC object are not really API-stable. (Really, we need accessor functions, so that people don't look at the LMIC object contents directly... but that's a separate topic.)

**Reply by Junaid-Q** [reply]
> * **passing in the** `rps`

Dear Mr. Moore,

I am really thankful to you for providing your kind suggestions. My sincere apology, I am still having a problem and couldn't sort it out. I thoroughly checked `LMIC.rps`  where all the information is given. However, I am quite confused with your second statement **to get the spreading factor, you use** `getSf()` **from** `lorabase.h` **passing in the `rps` value grabbed from** `LMIC.rps`. Would you be kind enough to please elaborate on how to pass in the `rps`?  

**Regarding the Power:**

I printed out `LMIC.radio_txpow` after `EV_TXCOMPLETE`. However, it gives me a constant value for instance 14 dBi. My concern is to see the real-time power consumption on the serial console as the packet is successfully transmitted to the gateway and the server.  I am using PicoScope 2203 and it give me the same result when I manually change the power and SF value in this line  `  LMIC_setDrTxpow(DR_SF7,14);`.


Technically it should give me different results if I put DR_SF8 and Power 20 instead of DR_SF7, 14.  I would really appreciate it if you could allow me for a video lecture as you had before on meetup. It was a really helpful lecture.


![image](https://user-images.githubusercontent.com/12583457/181808278-362c4fe6-6cd7-4a8a-8209-50e7876fd890.png)
![image](https://user-images.githubusercontent.com/12583457/181808391-06d534ee-df85-4b01-8f9e-e49467cf0a93.png)



For example, the behavior of the plot should be changed if we put different power and SF in the code. 
![image](https://user-images.githubusercontent.com/12583457/181810062-dc3b2b27-9a6f-4f3b-9b0e-7a528ca8a939.png)

**Comment by terrillmoore** [comment]
As this isn't really a bug, converting to a discussion.
